### PR TITLE
Frrist/deal state changes

### PR DESCRIPTION
**Pull request by frrist**
### Motivation
detect changes in the market deal proposal and market deal state
amts using generic adt array diff added in #2321

### CODE REVIEW

File: chain/events/state/predicates.go [hunk 1]
```diff
@@ -109,31 +113,189 @@ func (sp *StatePredicates) OnDealStateChanged(diffDealStates DiffDealStatesFunc)
...
+// - Modified Deals
+// - Removed Deals
+func (sp *StatePredicates) OnDealStateAmtChanged() DiffAdtArraysFunc {
+	return func(ctx context.Context, oldDealStates, newDealStates *adt.Array) (changed bool, user UserData, err error) {
+		dealStateChanges := new(MarketDealStateChanges)
+		if err := DiffAdtArray(oldDealStates, newDealStates, dealStateChanges); err != nil {
+			return false, nil, err
+		}
+
+		// proposals do not have a modified field since they are static.
```
Comment (dirkmc) [inline chain/events/state/predicates.go:258]: Does this comment belong on line 188?

Comment (dirkmc) [inline chain/events/state/predicates.go:258]: Does this comment belong on line 188?

File: chain/events/state/predicates.go [hunk 2]
```diff
@@ -109,31 +113,189 @@ func (sp *StatePredicates) OnDealStateChanged(diffDealStates DiffDealStatesFunc)
...
 // DealStateChangedForIDs detects changes in the deal state AMT for the given deal IDs
-func (sp *StatePredicates) DealStateChangedForIDs(dealIds []abi.DealID) DiffDealStatesFunc {
-	return func(ctx context.Context, oldDealStateRoot *amt.Root, newDealStateRoot *amt.Root) (changed bool, user UserData, err error) {
+func (sp *StatePredicates) DealStateChangedForIDs(dealIds []abi.DealID) DiffAdtArraysFunc {
+	return func(ctx context.Context, oldDealStateArray, newDealStateArray *adt.Array) (changed bool, user UserData, err error) {
 		changedDeals := make(ChangedDeals)
 		for _, dealID := range dealIds {
 			var oldDeal, newDeal market.DealState
-			err := oldDealStateRoot.Get(ctx, uint64(dealID), &oldDeal)
+			found, err := oldDealStateArray.Get(uint64(dealID), &oldDeal)
```
Comment (dirkmc) [inline chain/events/state/predicates.go:283]: The way this works has changed on the `next` branch:
https://github.com/filecoin-project/lotus/blob/next/chain/events/state/predicates.go#L127

File: chain/events/state/predicates.go (hunk 1 above)
Comment (frrist) [reply chain/events/state/predicates.go:258]: doh, you are right. will address.

File: chain/events/state/predicates.go (hunk 2 above)
Comment (frrist) [reply chain/events/state/predicates.go:283]: The case when a deal cannot be found is handled on the lines below. The `adt` package handles this case correctly I believe: https://github.com/filecoin-project/specs-actors/blob/3cc8bf59f1b983da1077277bd02a65f2c93eb7b6/actors/util/adt/array.go#L103

Comment (dirkmc) [reply chain/events/state/predicates.go:283]: On the `next` branch, when a deal is not found the code still adds an entry to the changedDeals map for that dealID with `nil` for `From` / `To`

Comment (frrist) [reply chain/events/state/predicates.go:283]: Ahh I see now and will address, thanks.

**Pull request merged**